content/develop/concepts/architecture/fragments

```diff
@@ -0,0 +1,143 @@
+
+## Use cases for fragments
+
+Fragments are versatile and applicable to a wide variety of circumstances. Here are just a few, common scenarios where fragments are useful:
+
+- Your app has multiple visualizations and each one takes time to load, but you have a filter input that only updates one of them.
+- You have a dynamic form that doesn't need to update the rest of your app (until the form is complete).
+- You want to automatically update a single component or group of components to stream data.
+
+## Defining and calling a fragment
+
+Streamlit provides a decorator to turn any function into a fragment function. When you call a fragment function that contains a widget, a user triggers a _partial rerun_ instead of a full rerun when they interact with that fragment's widget. During a partial rerun, your fragment function is re-executed. Anything within the main body of your fragment is updated on the frontend while the rest of your app remains the same. We will discuss what happens when you write to other containers within a fragment later on.
+
+Here is a basic example of defining and calling a fragment function:
+
+```ptyhon
+import streamlit as st
+
+@st.experimental_fragment
+def fragment_function():
+    if st.button("Hi!"):
+        st.write("Hi back!")
+
+fragment_function()
+```
+
+If you want the main body of your fragment to appear in the sidebar or another container, call your fragment function inside a context manager.
+
+```python
+with st.sidebar:
+    fragment_function()
+```
+
+### Partial rerun execution flow
+
+Consider the following code with the explanation and diagram that follow.
+
+```python
+import streamlit as st
+
+st.title("My Awesome App")
+
+@st.experimental_fragment()
+def fragment_one():
+    cols = st.columns(2)
+    cols[0].toggle("Toggle")
+    cols[1].text_area("Enter text")
+
+@st.experimental_fragment()
+def fragment_two():
+    cols = st.columns(2)
+    cols[0].checkbox("Filter")
+    cols[1].file_uploader("Upload image")
+
+fragment_one()
+cols = st.columns(2)
+cols[0].selectbox("Select", [1,2,3], None)
+cols[1].button("Update")
+fragment_two()
+```
+
+When a user interacts with an input widget created inside a fragment, only the fragment reruns instead of rerunning the full script as usual. When a user interacts with an input widget created outside a fragment, the full script reruns as usual.
+
+If you run the code above, the full script will run top to bottom on your app's initial load. If you flip the toggle button, the first fragment will rerun which will redraw the toggle and text area in the app while leaving everything else unchanged. If you then click the checkbox, the second fragment will rerun and consequently redraw the checkbox and file uploader. Everything else remains unchanged. Finally, if you click the update button, the full script will rerun and Streamlit will redraw everything.
+
+![Diagram of fragment execution flow](/images/concepts/fragment_diagram.png)
```

Comment:
Here's the source file for this image, in case anyone wants to tweak it.
- Download the file and remove the `.json` extension
- Navigate to excalidraw.com and do Open -> Load from File

[fragment_diagram.excalidraw.json](https://github.com/streamlit/docs/files/14887660/fragment_diagram.excalidraw.json)

Reply:
Awesome. I wanted to tweak it in a follow up, so this is great!